README.rst
.. image:: https://travis-ci.org/scanny/python-pptx.svg?branch=master
   :target: https://travis-ci.org/scanny/python-pptx

*python-pptx* is a Python library for creating and updating PowerPoint (.pptx)
files.

A typical use would be generating a customized PowerPoint presentation from
database content, downloadable by clicking a link in a web application.
Several developers have used it to automate production of presentation-ready
engineering status reports based on information held in their work management
system. It could also be used for making bulk updates to a library of
presentations or simply to automate the production of a slide or two that
would be tedious to get right by hand.

More information is available in the `python-pptx documentation`_.

Browse `examples with screenshots`_ to get a quick idea what you can do with
python-pptx.

.. _`python-pptx documentation`:
   https://python-pptx.readthedocs.org/en/latest/

.. _`examples with screenshots`:
   https://python-pptx.readthedocs.org/en/latest/user/quickstart.html


HISTORY.rst
.. :changelog:

Release History
---------------

0.6.18 (2019-05-02)
+++++++++++++++++++

- .text property getters encode line-break as a vertical-tab (VT, '\v', ASCII 11/x0B).
  This is consistent with PowerPoint's copy/paste behavior and allows like-breaks (soft
  carriage-return) to be distinguished from paragraph boundary. Previously, a line-break
  was encoded as a newline ('\n') and was not distinguishable from a paragraph boundary.

  .text properties include Shape.text, _Cell.text, TextFrame.text, _Paragraph.text and
  _Run.text.

- .text property setters accept vertical-tab character and place a line-break element in
  that location. All other control characters other than horizontal-tab ('\t') and
  newline ('\n') in range \x00-\x1F are accepted and escaped with plain-text like
  "_x001B" for ESC (ASCII 27).

  Previously a control character other than tab or newline in an assigned string would
  trigger an exception related to invalid XML character.


0.6.17 (2018-12-16)
+++++++++++++++++++

- Add SlideLayouts.remove() - Delete unused slide-layout
- Add SlideLayout.used_by_slides - Get slides based on this slide-layout
- Add SlideLayouts.index() - Get index of slide-layout in master
- Add SlideLayouts.get_by_name() - Get slide-layout by its str name


0.6.16 (2018-11-09)
+++++++++++++++++++

- Feature #395 DataLabels.show_* properties, e.g. .show_percentage
- Feature #453 Chart data tolerates None for labels


0.6.15 (2018-09-24)
+++++++++++++++++++

- Fix #436 ValueAxis._cross_xAx fails on c:dateAxis


0.6.14 (2018-09-24)
+++++++++++++++++++

- Add _Cell.merge()
- Add _Cell.split()
- Add _Cell.__eq__()
- Add _Cell.is_merge_origin
- Add _Cell.is_spanned
- Add _Cell.span_height
- Add _Cell.span_width
- Add _Cell.text getter
- Add Table.iter_cells()
- Move pptx.shapes.table module to pptx.table
- Add user documentation 'Working with tables'


0.6.13 (2018-09-10)
+++++++++++++++++++

- Add Chart.font
- Fix #293 Can't hide title of single-series Chart
- Fix shape.width value is not type Emu
- Fix add a:defRPr with c:rich (fixes some font inheritance breakage)


0.6.12 (2018-08-11)
+++++++++++++++++++

- Add Picture.auto_shape_type
- Remove Python 2.6 testing from build
- Update dependencies to avoid vulnerable Pillow version
- Fix #260, #301, #382, #401
- Add _Paragraph.add_line_break()
- Add Connector.line


0.6.11 (2018-07-25)
+++++++++++++++++++

- Add gradient fill.
- Add experimental "turbo-add" option for producing large shape-count slides.


0.6.10 (2018-06-11)
+++++++++++++++++++

- Add `shape.shadow` property to autoshape, connector, picture, and group
  shape, returning a `ShadowFormat` object.
- Add `ShadowFormat` object with read/write (boolean) `.inherit` property.
- Fix #328 add support for 26+ series in a chart


0.6.9 (2018-05-08)
++++++++++++++++++

- Add `Picture.crop_x` setters, allowing picture cropping values to be set,
  in addition to interrogated.
- Add `Slide.background` and `SlideMaster.background`, allowing the
  background fill to be set for an individual slide or for all slides based
  on a slide master.
- Add option `shapes` parameter to `Shapes.add_group_shape`, allowing a group
  shape to be formed from a number of existing shapes.
- Improve efficiency of `Shapes._next_shape_id` property to improve
  performance on high shape-count slides.


0.6.8 (2018-04-18)
++++++++++++++++++

- Add `GroupShape`, providing properties specific to a group shape, including
  its `shapes` property.
- Add `GroupShapes`, providing access to shapes contained in a group shape.
- Add `SlideShapes.add_group_shape()`, allowing a group shape to be added to
  a slide.
- Add `GroupShapes.add_group_shape()`, allowing a group shape to be added to
  a group shape, enabling recursive, multi-level groups.
- Add support for adding jump-to-named-slide behavior to shape and run
  hyperlinks.


0.6.7 (2017-10-30)
++++++++++++++++++

- Add `SlideShapes.build_freeform()`, allowing freeform shapes (such as maps)
  to be specified and added to a slide.
- Add support for patterned fills.
- Add `LineFormat.dash_style` to allow interrogation and setting of dashed
  line styles.


0.6.6 (2017-06-17)
++++++++++++++++++

- Add `SlideShapes.add_movie()`, allowing video media to be added to a slide.

- fix #190 Accommodate non-conforming part names having '00' index segment.
- fix #273 Accommodate non-conforming part names having no index segment.
- fix #277 ASCII/Unicode error on non-ASCII multi-level category names
- fix #279 BaseShape.id warning appearing on placeholder access.


0.6.5 (2017-03-21)
++++++++++++++++++

- #267 compensate for non-conforming PowerPoint behavior on c:overlay element

- compensate for non-conforming (to spec) PowerPoint behavior related to
  c:dLbl/c:tx that results in "can't save" error when explicit data labels
  are added to bubbles on a bubble chart.


0.6.4 (2017-03-17)
++++++++++++++++++

- add Chart.chart_title and ChartTitle object
- #263 Use Number type to test for numeric category


0.6.3 (2017-02-28)
++++++++++++++++++

- add DataLabel.font
- add Axis.axis_title


0.6.2 (2017-01-03)
++++++++++++++++++

- add support for NotesSlide (slide notes, aka. notes page)
- add support for arbitrary series ordering in XML
- add Plot.categories providing access to hierarchical categories in an
  existing chart.
- add support for date axes on category charts, including writing a dateAx
  element for the category axis when ChartData categories are date or
  datetime.

**BACKWARD INCOMPATIBILITIES:**

Some changes were made to the boilerplate XML used to create new charts. This
was done to more closely adhere to the settings PowerPoint uses when creating
a chart using the UI. This may result in some appearance changes in charts
after upgrading. In particular:


pptx/enum/dml.py
# encoding: utf-8

"""Enumerations used by DrawingML objects."""

from __future__ import absolute_import, division, print_function, unicode_literals

from .base import (
    alias,
    Enumeration,
    EnumMember,
    ReturnValueOnlyEnumMember,
    XmlEnumeration,
    XmlMappedEnumMember,
)


class MSO_COLOR_TYPE(Enumeration):
    """
    Specifies the color specification scheme

    Example::

        from pptx.enum.dml import MSO_COLOR_TYPE

        assert shape.fill.fore_color.type == MSO_COLOR_TYPE.SCHEME
    """

    __ms_name__ = "MsoColorType"

    __url__ = (
        "http://msdn.microsoft.com/en-us/library/office/ff864912(v=office.15" ").aspx"
    )

    __members__ = (
        EnumMember("RGB", 1, "Color is specified by an |RGBColor| value"),
        EnumMember("SCHEME", 2, "Color is one of the preset theme colors"),
        EnumMember(
            "HSL",
            101,
            """
            Color is specified using Hue, Saturation, and Luminosity values
            """,
        ),
        EnumMember(
            "PRESET",
            102,
            """
            Color is specified using a named built-in color
            """,
        ),
        EnumMember(
            "SCRGB",
            103,
            """
            Color is an scRGB color, a wide color gamut RGB color space
            """,
        ),
        EnumMember(
            "SYSTEM",
            104,
            """
            Color is one specified by the operating system, such as the
            window background color.
            """,
        ),
    )


@alias("MSO_FILL")
class MSO_FILL_TYPE(Enumeration):
    """
    Specifies the type of bitmap used for the fill of a shape.

    Alias: ``MSO_FILL``

    Example::

        from pptx.enum.dml import MSO_FILL

        assert shape.fill.type == MSO_FILL.SOLID
    """

    __ms_name__ = "MsoFillType"

    __url__ = "http://msdn.microsoft.com/EN-US/library/office/ff861408.aspx"

    __members__ = (
        EnumMember(
            "BACKGROUND",
            5,
            """
            The shape is transparent, such that whatever is behind the shape
            shows through. Often this is the slide background, but if
            a visible shape is behind, that will show through.
            """,
        ),
        EnumMember("GRADIENT", 3, "Shape is filled with a gradient"),
        EnumMember(
            "GROUP",
            101,
            "Shape is part of a group and should inherit the "
            "fill properties of the group.",
        ),
        EnumMember("PATTERNED", 2, "Shape is filled with a pattern"),
        EnumMember("PICTURE", 6, "Shape is filled with a bitmapped image"),
        EnumMember("SOLID", 1, "Shape is filled with a solid color"),
        EnumMember("TEXTURED", 4, "Shape is filled with a texture"),
    )


@alias("MSO_LINE")
class MSO_LINE_DASH_STYLE(XmlEnumeration):
    """Specifies the dash style for a line.

    Alias: ``MSO_LINE``

    Example::

        from pptx.enum.dml import MSO_LINE

        shape.line.dash_style == MSO_LINE.DASH_DOT_DOT
    """

    __ms_name__ = "MsoLineDashStyle"

    __url__ = (
        "https://msdn.microsoft.com/en-us/vba/office-shared-vba/articles/mso"
        "linedashstyle-enumeration-office"
    )

    __members__ = (
        XmlMappedEnumMember("DASH", 4, "dash", "Line consists of dashes only."),
        XmlMappedEnumMember("DASH_DOT", 5, "dashDot", "Line is a dash-dot pattern."),
        XmlMappedEnumMember(
            "DASH_DOT_DOT", 6, "lgDashDotDot", "Line is a dash-dot-dot patte" "rn."
        ),
        XmlMappedEnumMember("LONG_DASH", 7, "lgDash", "Line consists of long dashes."),
        XmlMappedEnumMember(
            "LONG_DASH_DOT", 8, "lgDashDot", "Line is a long dash-dot patter" "n."
        ),
        XmlMappedEnumMember("ROUND_DOT", 3, "dot", "Line is made up of round dots."),
        XmlMappedEnumMember("SOLID", 1, "solid", "Line is solid."),
        XmlMappedEnumMember(
            "SQUARE_DOT", 2, "sysDash", "Line is made up of square dots."
        ),
        ReturnValueOnlyEnumMember("DASH_STYLE_MIXED", -2, "Not supported."),
    )


@alias("MSO_PATTERN")
class MSO_PATTERN_TYPE(XmlEnumeration):
    """Specifies the fill pattern used in a shape.

    Alias: ``MSO_PATTERN``

    Example::

        from pptx.enum.dml import MSO_PATTERN

        fill = shape.fill
        fill.patterned()
        fill.pattern == MSO_PATTERN.WAVE
    """

    __ms_name__ = "MsoPatternType"

    __url__ = (
        "https://msdn.microsoft.com/VBA/Office-Shared-VBA/articles/msopatter"
        "ntype-enumeration-office"
    )

    __members__ = (
        XmlMappedEnumMember("CROSS", 51, "cross", "Cross"),
        XmlMappedEnumMember(
            "DARK_DOWNWARD_DIAGONAL", 15, "dkDnDiag", "Dark Downward Diagona" "l"
        ),
        XmlMappedEnumMember("DARK_HORIZONTAL", 13, "dkHorz", "Dark Horizontal"),
        XmlMappedEnumMember(
            "DARK_UPWARD_DIAGONAL", 16, "dkUpDiag", "Dark Upward Diagonal"
        ),
        XmlMappedEnumMember("DARK_VERTICAL", 14, "dkVert", "Dark Vertical"),
        XmlMappedEnumMember(
            "DASHED_DOWNWARD_DIAGONAL", 28, "dashDnDiag", "Dashed Downward D" "iagonal"
        ),
        XmlMappedEnumMember("DASHED_HORIZONTAL", 32, "dashHorz", "Dashed Horizontal"),
        XmlMappedEnumMember(
            "DASHED_UPWARD_DIAGONAL", 27, "dashUpDiag", "Dashed Upward Diago" "nal"
        ),
        XmlMappedEnumMember("DASHED_VERTICAL", 31, "dashVert", "Dashed Vertical"),
        XmlMappedEnumMember("DIAGONAL_BRICK", 40, "diagBrick", "Diagonal Brick"),
        XmlMappedEnumMember("DIAGONAL_CROSS", 54, "diagCross", "Diagonal Cross"),
        XmlMappedEnumMember("DIVOT", 46, "divot", "Pattern Divot"),
        XmlMappedEnumMember("DOTTED_DIAMOND", 24, "dotDmnd", "Dotted Diamond"),
        XmlMappedEnumMember("DOTTED_GRID", 45, "dotGrid", "Dotted Grid"),
        XmlMappedEnumMember("DOWNWARD_DIAGONAL", 52, "dnDiag", "Downward Diagonal"),
        XmlMappedEnumMember("HORIZONTAL", 49, "horz", "Horizontal"),
        XmlMappedEnumMember("HORIZONTAL_BRICK", 35, "horzBrick", "Horizontal Brick"),
        XmlMappedEnumMember(
            "LARGE_CHECKER_BOARD", 36, "lgCheck", "Large Checker Board"
        ),
        XmlMappedEnumMember("LARGE_CONFETTI", 33, "lgConfetti", "Large Confetti"),
        XmlMappedEnumMember("LARGE_GRID", 34, "lgGrid", "Large Grid"),
        XmlMappedEnumMember(
            "LIGHT_DOWNWARD_DIAGONAL", 21, "ltDnDiag", "Light Downward Diago" "nal"
        ),
        XmlMappedEnumMember("LIGHT_HORIZONTAL", 19, "ltHorz", "Light Horizontal"),
        XmlMappedEnumMember(
            "LIGHT_UPWARD_DIAGONAL", 22, "ltUpDiag", "Light Upward Diagonal"
        ),
        XmlMappedEnumMember("LIGHT_VERTICAL", 20, "ltVert", "Light Vertical"),
        XmlMappedEnumMember("NARROW_HORIZONTAL", 30, "narHorz", "Narrow Horizontal"),
        XmlMappedEnumMember("NARROW_VERTICAL", 29, "narVert", "Narrow Vertical"),
        XmlMappedEnumMember("OUTLINED_DIAMOND", 41, "openDmnd", "Outlined Diamond"),
        XmlMappedEnumMember("PERCENT_10", 2, "pct10", "10% of the foreground color."),
        XmlMappedEnumMember("PERCENT_20", 3, "pct20", "20% of the foreground color."),
        XmlMappedEnumMember("PERCENT_25", 4, "pct25", "25% of the foreground color."),
        XmlMappedEnumMember("PERCENT_30", 5, "pct30", "30% of the foreground color."),
        XmlMappedEnumMember("PERCENT_40", 6, "pct40", "40% of the foreground color."),
        XmlMappedEnumMember("PERCENT_5", 1, "pct5", "5% of the foreground color."),
        XmlMappedEnumMember("PERCENT_50", 7, "pct50", "50% of the foreground color."),
        XmlMappedEnumMember("PERCENT_60", 8, "pct60", "60% of the foreground color."),
        XmlMappedEnumMember("PERCENT_70", 9, "pct70", "70% of the foreground color."),
        XmlMappedEnumMember("PERCENT_75", 10, "pct75", "75% of the foreground color."),
        XmlMappedEnumMember("PERCENT_80", 11, "pct80", "80% of the foreground color."),
        XmlMappedEnumMember("PERCENT_90", 12, "pct90", "90% of the foreground color."),
        XmlMappedEnumMember("PLAID", 42, "plaid", "Plaid"),
        XmlMappedEnumMember("SHINGLE", 47, "shingle", "Shingle"),
        XmlMappedEnumMember(
            "SMALL_CHECKER_BOARD", 17, "smCheck", "Small Checker Board"
        ),
        XmlMappedEnumMember("SMALL_CONFETTI", 37, "smConfetti", "Small Confetti"),
        XmlMappedEnumMember("SMALL_GRID", 23, "smGrid", "Small Grid"),
        XmlMappedEnumMember("SOLID_DIAMOND", 39, "solidDmnd", "Solid Diamond"),
        XmlMappedEnumMember("SPHERE", 43, "sphere", "Sphere"),
        XmlMappedEnumMember("TRELLIS", 18, "trellis", "Trellis"),
        XmlMappedEnumMember("UPWARD_DIAGONAL", 53, "upDiag", "Upward Diagonal"),
        XmlMappedEnumMember("VERTICAL", 50, "vert", "Vertical"),
        XmlMappedEnumMember("WAVE", 48, "wave", "Wave"),
        XmlMappedEnumMember("WEAVE", 44, "weave", "Weave"),
        XmlMappedEnumMember(
            "WIDE_DOWNWARD_DIAGONAL", 25, "wdDnDiag", "Wide Downward Diagona" "l"
        ),
        XmlMappedEnumMember(
            "WIDE_UPWARD_DIAGONAL", 26, "wdUpDiag", "Wide Upward Diagonal"
        ),
        XmlMappedEnumMember("ZIG_ZAG", 38, "zigZag", "Zig Zag"),
        ReturnValueOnlyEnumMember("MIXED", -2, "Mixed pattern."),
    )


@alias("MSO_THEME_COLOR")
class MSO_THEME_COLOR_INDEX(XmlEnumeration):
    """
    Indicates the Office theme color, one of those shown in the color gallery
    on the formatting ribbon.

    Alias: ``MSO_THEME_COLOR``

    Example::

        from pptx.enum.dml import MSO_THEME_COLOR

        shape.fill.solid()
        shape.fill.fore_color.theme_color == MSO_THEME_COLOR.ACCENT_1
    """

    __ms_name__ = "MsoThemeColorIndex"

    __url__ = (
        "http://msdn.microsoft.com/en-us/library/office/ff860782(v=office.15" ").aspx"
    )

    __members__ = (
        EnumMember("NOT_THEME_COLOR", 0, "Indicates the color is not a theme color."),
        XmlMappedEnumMember(
            "ACCENT_1", 5, "accent1", "Specifies the Accent 1 theme color."
        ),
        XmlMappedEnumMember(
            "ACCENT_2", 6, "accent2", "Specifies the Accent 2 theme color."
        ),
        XmlMappedEnumMember(
            "ACCENT_3", 7, "accent3", "Specifies the Accent 3 theme color."
        ),
        XmlMappedEnumMember(
            "ACCENT_4", 8, "accent4", "Specifies the Accent 4 theme color."
        ),
        XmlMappedEnumMember(
            "ACCENT_5", 9, "accent5", "Specifies the Accent 5 theme color."
        ),
        XmlMappedEnumMember(
            "ACCENT_6", 10, "accent6", "Specifies the Accent 6 theme color."
        ),
        XmlMappedEnumMember(
            "BACKGROUND_1", 14, "bg1", "Specifies the Background 1 theme " "color."
        ),
        XmlMappedEnumMember(
            "BACKGROUND_2", 16, "bg2", "Specifies the Background 2 theme " "color."
        ),
        XmlMappedEnumMember("DARK_1", 1, "dk1", "Specifies the Dark 1 theme color."),
        XmlMappedEnumMember("DARK_2", 3, "dk2", "Specifies the Dark 2 theme color."),
        XmlMappedEnumMember(
            "FOLLOWED_HYPERLINK",
            12,
            "folHlink",
            "Specifies the theme color" " for a clicked hyperlink.",
        ),
        XmlMappedEnumMember(
            "HYPERLINK", 11, "hlink", "Specifies the theme color for a hyper" "link."
        ),
        XmlMappedEnumMember("LIGHT_1", 2, "lt1", "Specifies the Light 1 theme color."),
        XmlMappedEnumMember("LIGHT_2", 4, "lt2", "Specifies the Light 2 theme color."),
        XmlMappedEnumMember("TEXT_1", 13, "tx1", "Specifies the Text 1 theme color."),
        XmlMappedEnumMember("TEXT_2", 15, "tx2", "Specifies the Text 2 theme color."),
        ReturnValueOnlyEnumMember(
            "MIXED",
            -2,
            "Indicates multiple theme colors are used, such as " "in a group shape.",
        ),
    )
